File: src/day01.pl
:- module(day01, [parse_input/2, part1/2, part2/2, solve/3,
                  k_sum/4]).

% Day 1: Report Repair.
% Input is one integer expense entry per line. Find the two (part 1) and
% then three (part 2) entries that sum to 2020; the answer is their product.

% parse_input(+Raw, -Entries)
% Split on newlines, drop blanks, read each line as an integer.
parse_input(Raw, Entries) :-
    split_string(Raw, "\n", " \t\r", Lines0),
    exclude(=(""), Lines0, Lines),
    maplist(number_string, Entries, Lines).

% k_sum(+K, +Target, +Entries, -Combo)
% Combo is a K-element combination of Entries (order preserved) whose sum
% is exactly Target. Nondeterministic: backtracking enumerates every such
% combination. Entries should be ascending and non-negative so the
% `X =< Target` guard can prune doomed branches early.
k_sum(0, 0, _, []).
k_sum(K, Target, [X|Xs], [X|Combo]) :-
    K > 0,
    X =< Target,
    K1 is K - 1,
    T1 is Target - X,
    k_sum(K1, T1, Xs, Combo).
k_sum(K, Target, [_|Xs], Combo) :-
    K > 0,
    k_sum(K, Target, Xs, Combo).

% entry_product(+K, +Entries, -Product)
% Product of the first K-combination of Entries summing to 2020.
% msort/2 (not sort/2!) keeps duplicates, so a report containing 1010
% twice is still handled correctly.
entry_product(K, Entries, Product) :-
    msort(Entries, Sorted),
    once(k_sum(K, 2020, Sorted, Combo)),
    foldl([X, Acc0, Acc]>>(Acc is Acc0 * X), Combo, 1, Product).

% part1(+Entries, -Answer)
part1(Entries, Answer) :-
    entry_product(2, Entries, Answer).

% part2(+Entries, -Answer)
part2(Entries, Answer) :-
    entry_product(3, Entries, Answer).

% solve(+Raw, -Part1, -Part2)
solve(Raw, Part1, Part2) :-
    parse_input(Raw, Entries),
    part1(Entries, Part1),
    part2(Entries, Part2).

File: python/day01.py
"""Day 1: Report Repair.

Input is one integer expense entry per line; find the 2 (part 1) / 3
(part 2) entries summing to 2020 and multiply them.  This is k-SUM, done
as a depth-first take/skip walk rather than itertools.combinations, so the
ascending-order bound can prune whole subtrees.
"""

import math
from pathlib import Path
from typing import Iterator

INPUT = Path(__file__).resolve().parent.parent / "inputs" / "day01.txt"


def parse_input(raw: str) -> list[int]:
    return [int(line) for line in raw.splitlines() if line.strip()]


def k_sum(k: int, target: int, entries: list[int], start: int = 0) -> Iterator[list[int]]:
    """Yield every k-element combination of entries[start:] summing to target.

    Entries must be sorted ascending and non-negative: that is what makes
    the `x > target` break sound.  Once one candidate overshoots, every
    later one does too, so the rest of the level can be abandoned.
    """
    if k == 0:
        if target == 0:
            yield []
        return
    for i in range(start, len(entries)):
        x = entries[i]
        if x > target:
            break  # ascending: everything after is too big as well
        for combo in k_sum(k - 1, target - x, entries, i + 1):
            yield [x, *combo]


def entry_product(k: int, entries: list[int]) -> int:
    combo = next(k_sum(k, 2020, sorted(entries)))
    return math.prod(combo)


def part1(entries: list[int]) -> int:
    return entry_product(2, entries)


def part2(entries: list[int]) -> int:
    return entry_product(3, entries)


def solve(raw: str) -> tuple[int, int]:
    entries = parse_input(raw)
    return part1(entries), part2(entries)


def main() -> None:
    p1, p2 = solve(INPUT.read_text())
    print(f"part1={p1} part2={p2}")


if __name__ == "__main__":
    raise SystemExit(main())
